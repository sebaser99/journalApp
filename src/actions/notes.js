import { db } from "../firebase/firebase-config";
import { addDoc, collection, doc, updateDoc, deleteDoc } from "firebase/firestore";
import { types } from "../types/types";
import { loadNotes } from "../helpers/loadNotes";
import Swal from "sweetalert2";
import { fileUpload } from "../helpers/fileUpload";

export const startNewNote = ()=>{
    return async(dispatch, getState) =>{
        const uid = getState().auth.uid;
        const newNote = {
            title: '',
            body: '',
            date: new Date().getTime(),
            url: ''
        }
        try{
            const docRef = await addDoc(collection(db, `${uid}/journal/notes`), newNote)
            console.log("Document written with ID: ", docRef.id)
            dispatch(activeNote(docRef.id, newNote))
            dispatch(addNewNote(docRef.id, newNote))
        }
        catch(e){
            console.log("Error adding a document:" , e)
        }
    }
}

export const addNewNote = (id, note)=> ({
    type: types.notesAddNew,
    payload: {
        id,
        ...note
    }
})

export const activeNote = (id, note)=>(
    {
        type: types.notesActive,
        payload: {
            id,
            ...note
        }

    }
) 

export const startLoadingNotes = (uid) =>{
    return async (dispatch) =>{
        const notes = await loadNotes(uid)
        dispatch(setNotes(notes))
    }
}

export const setNotes = (notes)=>(
    {
        type: types.notesLoad,
        payload: notes
    }
)

export const startSaveNote = (note)=>{
    return async (dispatch, getState)=> {
        const {uid} = getState().auth
        if(!note.url){
            delete note.url
        }
        const noteToSave = {...note}
        delete noteToSave.id
        await updateDoc(doc(db, `${uid}/journal/notes/${note.id}`), noteToSave)

        dispatch(refreshNoteUpdated(note.id, noteToSave))
        Swal.fire('Saved', note.title, "success")
        
    }
}

export const refreshNoteUpdated = (id, note) => ({
    type: types.notesUpdated,
    payload: {
        id, 
        note: {
            id,
            ...note
        }
    }
})

export const startUploading = (file) => {
    return async(dispatch, getState)=> {
        const {active: activeNote} = getState().notes;
        Swal.fire({
            title: 'Uploading',
            text: 'Please wait....',
            didOpen: ()=> {
                Swal.showLoading()
            }
        })
       const urlLoad = await fileUpload(file)
       activeNote.url = urlLoad

       dispatch(startSaveNote(activeNote))
       Swal.close()
    }
}

export const startDeleting = (id)=> {
    return async(dispatch, getState)=>{
        
        const uid = getState().auth.uid
        console.log(uid)
        await deleteDoc(doc(db, `${uid}/journal/notes/${id}`))
        dispatch(deleteNote(id))
    }
}

export const deleteNote = (id)=> ({
    type: types.notesDelete,
    payload: id
})

export const cleaningLogout = () => ({
    type: types.notesLogoutCleaning

})

export const searchingNotes = (search) => ({
    type: types.notesSearch,
    payload: search
}) 