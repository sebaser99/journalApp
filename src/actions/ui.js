import { types } from "../types/types";

export const uiSetError = (message)=>({
    type: types.uiSetError,
    payload: {
        msg: message
    }
})

export const uiRemoveError = ()=>({
    type: types.uiRemoveError
    
})

export const startLoading = ()=>({
    type: types.uiStartLoading,
  
})
export const finishLoading = ()=>({
    type: types.uiFinishLoading,
   
})
