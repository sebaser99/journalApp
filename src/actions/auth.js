import { auth } from "../firebase/firebase-config";
import { GoogleAuthProvider, signInWithPopup, createUserWithEmailAndPassword, updateProfile, signInWithEmailAndPassword } from "firebase/auth";
import { types } from "../types/types";
import { finishLoading, startLoading } from "./ui";
import Swal from 'sweetalert2';
import { cleaningLogout } from "./notes";

export const startLoginEmailPassword = (email, password)=> {
    return(dispatch)=>{
        dispatch(startLoading())

        return signInWithEmailAndPassword(auth, email, password)
            .then(({user}) =>{
                dispatch(login(user.uid, user.displayName))
                dispatch(finishLoading())
            })
            .catch(error =>{
                dispatch(finishLoading())
                Swal.fire('Error', error.message, 'error')
            }) 
        
    }
}

export const createUserWithEmailPassword = (email, password, name) =>{
    return(dispatch)=>{
         return createUserWithEmailAndPassword(auth, email, password)
        .then(async({user}) =>{
            await updateProfile(auth.currentUser, {
                displayName: name
            })
             dispatch(login(user.uid, user.displayName))
        })
        .catch(error => {
            Swal.fire('Error', error.message, 'error')
            console.log(error.code)
            
        })
    }
}

export const startGoogleLogin = ()=>{
    const googleAuthProvider = new GoogleAuthProvider();
    return(dispatch)=>{
        signInWithPopup(auth, googleAuthProvider)
        .then(({user})=>{
           dispatch(
            login(user.uid, user.displayName)
           )
  
        })
        .catch(err => console.log(err))
    }
}

export const login = (uid, displayName) => (
    {
        type: types.login,
        payload: {
            uid,
            displayName
        }
    }
)

export const startLogout = ()=> {
    return(dispatch)=>{
        auth.signOut()
        dispatch(logout())
        dispatch(cleaningLogout())
    }
}

export const logout = ()=> (
    {
        type: types.logout
    }
)

