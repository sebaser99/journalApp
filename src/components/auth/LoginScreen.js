import { Link } from "react-router-dom";
import { useDispatch, useSelector } from "react-redux";
import { useForm } from "../../hooks/useForm";
import {  startGoogleLogin, startLoginEmailPassword } from "../../actions/auth";
import validator  from "validator";
import { uiRemoveError, uiSetError} from "../../actions/ui";

export const LoginScreen = () => {
    const dispatch = useDispatch();
    const {loading, msgError} = useSelector(state => state.ui)
    const [ formValue, handleChangeInput]= useForm({
        email: "alexrodriguez@gmail.com",
        password: '123456'
    })
    const {email, password} = formValue;

    const handleLogin = (e)=>{
        e.preventDefault()
        if(isFormValid()){
            dispatch(startLoginEmailPassword(email, password))
        
        }        
    }
    const handleLoginGoogle = ()=>{
        dispatch(startGoogleLogin())
    }
    const isFormValid = ()=>{
        if(!validator.isEmail(email)){
            dispatch(uiSetError('Email is not valid'))
            return false
        } else if(password.length <= 5){
            dispatch(uiSetError('El password must be at least 6 characters'))
            return false 
        }
        dispatch(uiRemoveError())
        return true
    }
    return (
        <div className="animate__animated animate__fadeIn animate__fast">
            <h3 className="auth__title mb-5">Login</h3>
            {
                msgError && (
                    <div className="auth__alert-error">{msgError}</div>
                )
            }
            <form onSubmit={handleLogin}>
                <input  type='text' 
                    placeholder="Email"
                    name="email"
                    className="auth__input"
                    autoComplete='off'
                    value={email}
                    onChange={handleChangeInput}
                />
                <input  type='password' 
                    placeholder="Password"
                    name="password"
                    className="auth__input"
                    value={password}
                    onChange={handleChangeInput}
                />
                <button className="btn btn-primary" type="submit" disabled={loading} >Login</button>
                
                <div className="auth__social-networks">
                    <p>Login with social media networks</p>
                    <div className="google-btn"
                            onClick={handleLoginGoogle}
                        >
                        <div className="google-icon-wrapper">
                            <img className="google-icon" src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg" alt="google button" />
                        </div>
                        <p className="btn-text">
                            <b>Sign in with google</b>
                        </p>
                    </div>
                </div>
                <Link className="auth__link" to='/auth/register' >
                    Create new account
                </Link>

            </form>
        </div>
    )
}
