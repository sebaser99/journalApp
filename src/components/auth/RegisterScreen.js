import { Link } from "react-router-dom";
import {useForm} from '../../hooks/useForm';
import validator from 'validator';
import { useDispatch, useSelector } from "react-redux";
import { uiRemoveError, uiSetError } from "../../actions/ui";
import { createUserWithEmailPassword } from "../../actions/auth";


export const RegisterScreen = () => {
    const dispatch = useDispatch();
    const {msgError} = useSelector(state => state.ui)
   
    const [ formValue, handleChangeInput]= useForm({
        name: 'Alex Rodriguez',
        email: "alexrodriguez@gmail.com",
        password: '123456',
        password2: '123456'
    })
    const {name, email, password, password2} = formValue

    const handleRegister = (e)=>{
        e.preventDefault()
        if(isFormValid()){
            dispatch(createUserWithEmailPassword(email, password, name))
        }
    }
    const isFormValid = ()=> {
        if(name.trim().length === 0){
            dispatch(uiSetError('Name is required'))
            return false
        } else if(!validator.isEmail(email)){
            dispatch(uiSetError('Email is not valid'))
            return false
        }else if(password !== password2 || password.length <= 5){
            dispatch(uiSetError('Password should be a least 6 characthers and match each other'))
            return false
        }
        dispatch(uiRemoveError())
        return true
    }
    return (
        <>
        <h3 className="auth__title mb-5">Register</h3>
        {msgError && (
            <div className="auth__alert-error">{msgError}</div>
        )}
        <form onSubmit={handleRegister}>
            <input  type='text' 
                placeholder="Name"
                name="name"
                className="auth__input"
                autoComplete='off'
                value={name}
                onChange={handleChangeInput}
            />
            <input  type='text' 
                placeholder="Email"
                name="email"
                className="auth__input"
                autoComplete='off'
                value={email}
                onChange={handleChangeInput}
            />
            <input  type='password' 
                placeholder="Password"
                name="password"
                className="auth__input"
                value={password}
                onChange={handleChangeInput}
            />
            <input  type='password' 
                placeholder="Confirm Password"
                name="password2"
                className="auth__input"
                value={password2}
                onChange={handleChangeInput}
            />
            <button className="btn btn-primary" type="submit">Register</button>

           
            <Link className="auth__link" to='/auth/login' >
                I have a account
            </Link>

        </form>
    </>
    )
}
