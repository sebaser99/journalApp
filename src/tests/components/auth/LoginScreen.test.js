import { mount } from "enzyme";
import React from 'react';
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";

import configureStore  from 'redux-mock-store';
import thunk from 'redux-thunk';
import '@testing-library/jest-dom';

import { LoginScreen } from "../../../components/auth/LoginScreen";
import { startGoogleLogin, startLoginEmailPassword } from "../../../actions/auth";

jest.mock("../../../actions/auth", ()=> ({
    startGoogleLogin: jest.fn(),
    startLoginEmailPassword: jest.fn()
}))
const middlewares = [thunk]
const mockStore = configureStore(middlewares)
const initState = {
    auth: {},
    ui: {
        loading: false,
        msgError: null
    }
}

let store = mockStore(initState)
store.dispatch = jest.fn()
const wrapper = mount(
    <Provider store={store}>
        <MemoryRouter>
            <LoginScreen />
        </MemoryRouter>
    </Provider>
    )


describe('pruebas en LoginScreen', ()=> {
    test('debe mostrar el componente correctamente', ()=> {
        expect(wrapper).toMatchSnapshot()
    })

    test('debe de disparar la acciÃ³n de startGoogleLogin', ()=> {
        wrapper.find('.google-btn').prop('onClick')()
        expect(startGoogleLogin).toHaveBeenCalled()
    }),
    test('debe de disparar la accion startLoginEmailPassword con los argumentos correctos', ()=> {
        wrapper.find('form').prop('onSubmit')({
            preventDefault(){}
        })
        expect(startLoginEmailPassword).toHaveBeenLastCalledWith('alexrodriguez@gmail.com', '123456')
    })
})