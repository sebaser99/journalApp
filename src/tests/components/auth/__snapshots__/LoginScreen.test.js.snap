// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en LoginScreen debe mostrar el componente correctamente 1`] = `
<div
  className="animate__animated animate__fadeIn animate__fast"
>
  <h3
    className="auth__title mb-5"
  >
    Login
  </h3>
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="auth__input"
      name="email"
      onChange={[Function]}
      placeholder="Email"
      type="text"
      value="alexrodriguez@gmail.com"
    />
    <input
      className="auth__input"
      name="password"
      onChange={[Function]}
      placeholder="Password"
      type="password"
      value="123456"
    />
    <button
      className="btn btn-primary"
      disabled={false}
      type="submit"
    >
      Login
    </button>
    <div
      className="auth__social-networks"
    >
      <p>
        Login with social media networks
      </p>
      <div
        className="google-btn"
        onClick={[Function]}
      >
        <div
          className="google-icon-wrapper"
        >
          <img
            alt="google button"
            className="google-icon"
            src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg"
          />
        </div>
        <p
          className="btn-text"
        >
          <b>
            Sign in with google
          </b>
        </p>
      </div>
    </div>
    <a
      className="auth__link"
      href="/auth/register"
      onClick={[Function]}
    >
      Create new account
    </a>
  </form>
</div>
`;
